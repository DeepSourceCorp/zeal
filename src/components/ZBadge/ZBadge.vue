<template>
  <div class="relative inline-block z-badge">
    <slot></slot>
    <sup
      v-show="!hidden && (content || content === 0 || isDot)"
      v-text="content"
      class="absolute inline-block text-xs text-center transform translate-x-full -translate-y-1/2 z-badge-content"
      :class="[
        getBgColor,
        getSize,
        getPosition,
        `text-${textColor}`,
        {
          'p-0 rounded-full': isDot,
          'px-2 leading-4 rounded-full': value
        }
      ]"
    >
    </sup>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'

export default Vue.extend({
  name: 'ZBadge',
  components: {},
  props: {
    value: [String, Number],
    max: Number,
    isDot: Boolean,
    hidden: Boolean,
    size: {
      type: String,
      default: 'sm',
      validator(val) {
        return ['sm', 'md', 'lg'].indexOf(val) > -1
      }
    },
    type: {
      type: String,
      default: 'default',
      validator(val) {
        return ['default', 'success', 'warning', 'info', 'danger'].indexOf(val) > -1
      }
    },
    position: {
      type: String
    },
    textColor: {
      type: String,
      default: 'vanilla-100'
    }
  },
  computed: {
    getBgColor() {
      const colors: Record<string, string> = {
        default: '',
        success: 'juniper',
        warning: 'honey',
        info: 'aqua',
        danger: 'cherry'
      }
      return `bg-${colors[this.type as string]}`
    },
    getSize() {
      const sizes: Record<string, string> = {
        sm: 'w-2 h-2',
        md: 'w-2.5 h-2.5',
        lg: 'w-3 h-3'
      }
      return this.isDot ? sizes[this.size as string] : ''
    },
    getPosition(): string {
      if (this.position) return this.position as string
      else {
        if (this.value) return 'top-0 right-3'
        if (this.isDot) return 'top-0 right-1'
        return 'top-0'
      }
    },
    content() {
      if (this.isDot) return
      const value = this.value,
        max = this.max
      if (typeof value === 'number' && typeof max === 'number') {
        return max < value ? `${max}+` : value
      }
      return value
    }
  }
})
</script>
