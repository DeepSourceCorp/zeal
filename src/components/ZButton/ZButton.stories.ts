import ZButton from './ZButton.vue'

import { action } from '@storybook/addon-actions'

import '../../assets/css/tailwind.css'
import '../../assets/css/typography.css'
import '../../assets/css/layout.css'

export default {
  title: 'Zeal/Button',
  component: ZButton,
  excludeStories: /.*Data$/
}

export const DefaultWithoutText = () => ({
  components: { ZButton },
  template: '<z-button></z-button>'
})

export const ButtonWithText = () => ({
  components: { ZButton },
  template: '<z-button color="primary">Hello World</z-button>'
})

export const Colors = () => ({
  components: { ZButton },
  template: `<div>
        <z-button color="primary">Hello World</z-button>
        <z-button color="secondary">Hello World</z-button>
    </div>`
})

export const Sizes = () => ({
  components: { ZButton },
  template: `<div>
        <z-button size="small">Small Button</z-button>
        <z-button size="medium">Medium Button</z-button>
        <z-button size="large">Large Button</z-button>
        <z-button size="xlarge">Extra Large Button</z-button>
    </div>`
})

export const States = () => ({
  components: { ZButton },
  template: `<div>
        <z-button active>Active Button</z-button>
        <z-button disabled>Disabled Button</z-button>
        <z-button color="secondary" active>Active Secondary Button</z-button>
        <z-button color="secondary" disabled>Disabled Secondary Button</z-button>
    </div>`
})

export const FullWidthButton = () => ({
  components: { ZButton },
  template: `<div>
        <z-button full-width>Full Width Button</z-button>
    </div>`
})

export const DefaultLinkButton = () => ({
  components: { ZButton },
  template: `<div>
        <z-button color="link" type="link" to="https://deepsource.io/">Link Button</z-button>
    </div>`
})

export const ActiveLinkButton = () => ({
  components: { ZButton },
  template: `<div>
        <z-button color="link" type="link" to="https://deepsource.io/" active>Link Button</z-button>
    </div>`
})

export const DisabledLinkButton = () => ({
  components: { ZButton },
  template: `<div>
        <z-button color="link" type="link" to="https://deepsource.io/" disabled>Link Button</z-button>
    </div>`
})
