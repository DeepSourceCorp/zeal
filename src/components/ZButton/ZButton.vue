<template>
  <component
    :is="type === 'link' || to ? 'a' : 'button'"
    :href="to"
    class="z-btn"
    v-bind="$attrs"
    :disabled="disabled"
    v-on="$listeners"
    :class="[
      `${color && `z-btn--${color}`}`,
      `${size && `z-btn--${size}`}`,
      {
        'is-disabled': this.disabled !== false,
        'is-active': this.active !== false,
        'full-width': this.fullWidth !== false
      },
      customClasses,
      `${!color && !customClasses ? 'z-btn--primary' : ''}`
    ]"
    :type="type"
  >
    <slot>Click</slot>
  </component>
</template>

<script>
export default {
  name: 'ZButton',
  props: {
    color: {
      default: '',
      type: String
    },
    type: {
      default: 'button',
      type: String
    },
    customClasses: {
      default: '',
      type: String
    },
    size: {
      default: '',
      type: String
    },
    disabled: {
      default: false,
      type: Boolean
    },
    active: {
      default: false,
      type: Boolean
    },
    fullWidth: {
      default: false,
      type: Boolean
    },
    to: {
      default: '',
      type: String
    }
  },
  methods: {
    handleClick(event) {
      this.$emit('click', event)
    }
  }
}
</script>

<style lang="css" scoped>
/* Common Button Style */

.z-btn {
  @apply inline-flex items-center px-6 text-base font-medium border-transparent h-10 rounded-sm relative;
}

.is-disabled {
  @apply opacity-50 cursor-not-allowed;
}

.full-width {
  @apply w-full inline-block;
}

/* Button - Primary Style */

.z-btn--primary {
  @apply bg-juniper hover:bg-light_juniper text-ink-400;
}

.z-btn--primary.is-disabled {
  @apply hover:bg-juniper;
}

/* Button - Secondary Style */

.z-btn--secondary {
  @apply bg-ink-300 hover:bg-ink-200 text-vanilla-100 border border-slate;
}

.z-btn--secondary.is-disabled {
  @apply hover:bg-ink-300;
}

.z-btn--secondary.is-active {
  @apply bg-ink-100;
}

.z-btn--link {
  @apply font-normal text-juniper hover:underline;
}

.z-btn--link.is-active {
  @apply underline;
}

.z-btn--link.is-disabled {
  @apply text-slate no-underline;
}

/* Different Sizes of the Button*/
.z-btn--small {
  @apply h-8 text-xs;
}

.z-btn--medium {
  @apply h-10 text-base;
}

.z-btn--large {
  @apply h-12 text-lg;
}

.z-btn--xlarge {
  @apply h-16 text-lg;
}
</style>
