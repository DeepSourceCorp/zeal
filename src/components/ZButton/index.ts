import ZButton from './ZButton.vue';

export { ZButton };
