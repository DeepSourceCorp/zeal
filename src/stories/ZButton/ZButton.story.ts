import ZButton from '../../components/ZButton/ZButton.vue'
import { Meta, Story } from '@storybook/addon-docs/blocks'
import { component } from 'vue/types/umd'

export const ButtonWithText = () => ({
  components: { ZButton },
  template: '<z-button>Hello Button</button>',
  title: 'ZButton'
})

ButtonWithText.story = {
  name: 'Button with Text'
}
