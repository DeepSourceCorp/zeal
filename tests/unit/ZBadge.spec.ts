import ZBadge from '../../src/components/ZBadge'
import ZButton from '../../src/components/ZButton'

import { shallowMount, mount } from '@vue/test-utils'

const slotData = {
  default: 'Badge'
}

const BasicBadge = {
  template: `<z-badge type="danger" :value="12">
            <z-button color="primary">Likes</z-button>
        </z-badge>`,
  components: {
    ZBadge,
    ZButton
  }
}

describe('Badge', () => {
  it('renders a badge as a dot', () => {
    const wrapper = shallowMount(ZBadge, {
      slots: slotData,
      propsData: {
        isDot: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders a badge with content', () => {
    const wrapper = mount(BasicBadge)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders with state', () => {
    expect(
      shallowMount(ZBadge, {
        propsData: {
          state: 'success'
        },
        slots: slotData
      }).html()
    ).toMatchSnapshot()

    expect(
      shallowMount(ZBadge, {
        propsData: {
          state: 'danger'
        },
        slots: slotData
      }).html()
    ).toMatchSnapshot()

    expect(
      shallowMount(ZBadge, {
        propsData: {
          state: 'warning'
        },
        slots: slotData
      }).html()
    ).toMatchSnapshot()

    expect(
      shallowMount(ZBadge, {
        propsData: {
          state: 'info'
        },
        slots: slotData
      }).html()
    ).toMatchSnapshot()

    expect(
      shallowMount(ZBadge, {
        propsData: {
          state: 'default'
        },
        slots: slotData
      }).html()
    ).toMatchSnapshot()
  })

  it('renders with size', () => {
    expect(
      shallowMount(ZBadge, {
        propsData: {
          size: 'sm'
        },
        slots: slotData
      }).html()
    ).toMatchSnapshot()

    expect(
      shallowMount(ZBadge, {
        propsData: {
          size: 'md'
        },
        slots: slotData
      }).html()
    ).toMatchSnapshot()

    expect(
      shallowMount(ZBadge, {
        propsData: {
          size: 'lg'
        },
        slots: slotData
      }).html()
    ).toMatchSnapshot()
  })
})
