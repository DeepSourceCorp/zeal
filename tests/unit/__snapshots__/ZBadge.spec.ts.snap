// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge renders a badge as a dot 1`] = `
<div class="relative inline-block z-badge">Badge <div class="text-center transform z-badge-content bg- w-2 h-2 top-px right-px text-vanilla-100 absolute translate-x-1/2 -translate-y-1/2 p-0 rounded-full"><span></span></div>
</div>
`;

exports[`Badge renders a badge with content 1`] = `
<div class="relative inline-block z-badge"><button href="#" type="button" class="inline-flex items-center font-medium transition-colors duration-300 ease-in-out rounded-sm z-btn focus:outline-none whitespace-nowrap text-ink-400 justify-center h-10 px-4 py-2 text-base space-x-2 leading-8 bg-juniper hover:bg-juniper-600">
    <!----> Likes
  </button>
  <div class="text-center transform z-badge-content bg-cherry top-px right-0 h-6 w-6 flex justify-center items-center text-xxs font-bold text-vanilla-100 absolute translate-x-1/2 -translate-y-1/2 leading-4 rounded-full"><span>12</span></div>
</div>
`;

exports[`Badge renders with size 1`] = `
<div class="relative inline-block z-badge">Badge <div class="text-center transform z-badge-content bg- top-0 inline-block text-xs px-2 text-vanilla-100 absolute translate-x-1/2 -translate-y-1/2" style="display: none;"><span></span></div>
</div>
`;

exports[`Badge renders with size 2`] = `
<div class="relative inline-block z-badge">Badge <div class="text-center transform z-badge-content bg- top-0 inline-block text-xs px-2 text-vanilla-100 absolute translate-x-1/2 -translate-y-1/2" style="display: none;"><span></span></div>
</div>
`;

exports[`Badge renders with size 3`] = `
<div class="relative inline-block z-badge">Badge <div class="text-center transform z-badge-content bg- top-0 inline-block text-xs px-2 text-vanilla-100 absolute translate-x-1/2 -translate-y-1/2" style="display: none;"><span></span></div>
</div>
`;

exports[`Badge renders with state 1`] = `
<div class="relative inline-block z-badge" state="success">Badge <div class="text-center transform z-badge-content bg- top-0 inline-block text-xs px-2 text-vanilla-100 absolute translate-x-1/2 -translate-y-1/2" style="display: none;"><span></span></div>
</div>
`;

exports[`Badge renders with state 2`] = `
<div class="relative inline-block z-badge" state="danger">Badge <div class="text-center transform z-badge-content bg- top-0 inline-block text-xs px-2 text-vanilla-100 absolute translate-x-1/2 -translate-y-1/2" style="display: none;"><span></span></div>
</div>
`;

exports[`Badge renders with state 3`] = `
<div class="relative inline-block z-badge" state="warning">Badge <div class="text-center transform z-badge-content bg- top-0 inline-block text-xs px-2 text-vanilla-100 absolute translate-x-1/2 -translate-y-1/2" style="display: none;"><span></span></div>
</div>
`;

exports[`Badge renders with state 4`] = `
<div class="relative inline-block z-badge" state="info">Badge <div class="text-center transform z-badge-content bg- top-0 inline-block text-xs px-2 text-vanilla-100 absolute translate-x-1/2 -translate-y-1/2" style="display: none;"><span></span></div>
</div>
`;

exports[`Badge renders with state 5`] = `
<div class="relative inline-block z-badge" state="default">Badge <div class="text-center transform z-badge-content bg- top-0 inline-block text-xs px-2 text-vanilla-100 absolute translate-x-1/2 -translate-y-1/2" style="display: none;"><span></span></div>
</div>
`;
